packages/next/src/client/app-index.tsx

```diff
@@ -33,6 +44,72 @@ window.addEventListener('error', (ev: WindowEventMap['error']): void => {
 
 const appElement: HTMLElement | Document | null = document
 
+function finishRunningAction(
+  actionQueue: AppRouterActionQueue,
+  setState: DispatchStatePromise
+) {
+  if (actionQueue.pending !== null) {
+    actionQueue.pending = actionQueue.pending.next
+    if (actionQueue.pending !== null) {
+      // eslint-disable-next-line @typescript-eslint/no-use-before-define
+      runAction(actionQueue, actionQueue.pending.payload, setState)
+    }
+  }
+}
+async function runAction(
+  actionQueue: AppRouterActionQueue,
+  payload: ReducerActions,
+  setState: DispatchStatePromise
+) {
+  const prevState = actionQueue.state
+  const promise = actionQueue.action(prevState, payload)
+
+  promise.then(
+    (nextState: AppRouterState | null) => {
+      actionQueue.state = nextState
+
+      if (actionQueue.devToolsInstance) {
+        actionQueue.devToolsInstance.send(payload, nextState)
+      }
+
+      finishRunningAction(actionQueue, setState)
+    },
+    () => {
+      finishRunningAction(actionQueue, setState)
+    }
+  )
+
+  startTransition(() => {
+    setState(promise)
+  })
+}
+
+function dispatchAction(
+  actionQueue: AppRouterActionQueue,
+  payload: ReducerActions,
+  setState: DispatchStatePromise
+) {
+  const newAction: ActionQueueNode = {
+    payload,
+    next: null,
+  }
+  // Check if the queue is empty or if the action is a navigation action
+  if (actionQueue.pending === null || payload.type === 'navigate') {
+    // The queue is empty, so add the action and start it immediately
+    // We also immediately start navigations as those shouldn't be blocked
+    actionQueue.pending = newAction
+    runAction(actionQueue, newAction.payload, setState)
+  } else {
+    // The queue is not empty, so add the action to the end of the queue
+    // It will be started by finishRunningAction after the previous action finishes
+    let last = actionQueue.pending
+    while (last.next !== null) {
```

Comment:
You should just track `first` and `last` separately so you don't have to loop through the whole thing every time. Another approach is to use a circular list, then first is the same as `last.next`.